admin-dev/themes/new-theme/js/pages/product/components/categories/category-tree-selector.ts

```diff
@@ -0,0 +1,471 @@
+    $.fancybox({
+      type: 'iframe',
+      width: '90%',
+      height: '90%',
+      fitToView: false,
+      autoSize: false,
+      content: modalContent.html(),
+      afterShow: () => {
+        this.initModal();
+      },
+    });
+  }
+
+  private async initModal(): Promise<void> {
+    this.modalContainer = document.querySelector(ProductCategoryMap.categoriesModalContainer) as HTMLElement;
+    this.categoryTree = this.modalContainer.querySelector(ProductCategoryMap.categoryTree) as HTMLElement;
+    this.prototypeTemplate = this.categoryTree.dataset.prototype ? this.categoryTree.dataset.prototype : null;
+    this.prototypeName = this.categoryTree.dataset.prototypeName ? this.categoryTree.dataset.prototypeName : null;
+    this.expandAllButton = this.modalContainer.querySelector(ProductCategoryMap.expandAllButton);
+    this.reduceAllButton = this.modalContainer.querySelector(ProductCategoryMap.reduceAllButton);
+    this.tags = new Tags(
+      this.eventEmitter,
+      `${ProductCategoryMap.categoriesModalContainer} ${ProductCategoryMap.tagsContainer}`,
+    );
+    this.tags.render(this.selectedCategories);
+    this.treeCategories = await getCategories();
+
+    this.initTypeaheadData(this.treeCategories, '');
+    this.initTypeahead();
+    this.initTree();
+    this.listenCancelChanges();
+    this.applyCategoryTreeChanges();
+    this.eventEmitter.on(ProductEventMap.categories.categoryRemoved, (categoryId) => this.unselectCategory(categoryId));
+  }
+
+  private applyCategoryTreeChanges(): void {
+    if (!this.modalContainer) {
+      return;
+    }
+
+    const applyBtn = this.modalContainer.querySelector(ProductCategoryMap.applyCategoriesBtn) as HTMLElement;
+
+    applyBtn.addEventListener('click', () => {
+      this.eventEmitter.emit(ProductEventMap.categories.applyCategoryTreeChanges, {
+        categories: this.selectedCategories,
+      });
+      this.closeModal();
+    });
+  }
+
+  private listenCancelChanges(): void {
+    if (!this.modalContainer) {
+      return;
+    }
+
+    const cancelBtn = this.modalContainer.querySelector(ProductCategoryMap.cancelCategoriesBtn) as HTMLElement;
+    cancelBtn.addEventListener('click', () => this.closeModal());
+  }
+
+  private initTree(): void {
+    const categoryTree = this.categoryTree;
+
+    if (!categoryTree) {
+      return;
+    }
+
+    this.treeCategories.forEach((treeCategory) => {
+      const treeCategoryElement = this.generateCategoryTree(treeCategory);
+      categoryTree.append(treeCategoryElement);
+    });
+
+    if (this.expandAllButton) {
+      this.expandAllButton.addEventListener('click', () => {
+        this.toggleAll(true);
+      });
+    }
+
+    if (this.reduceAllButton) {
+      this.reduceAllButton.addEventListener('click', () => {
+        this.toggleAll(false);
+      });
+    }
+
+    categoryTree.querySelectorAll(ProductCategoryMap.checkboxInput).forEach((checkbox) => {
+      if (checkbox instanceof HTMLInputElement) {
```

Comment:
It's a bit weird to check the type since the selector is supposed to match what we expect But why not ^^
However maybe it can be simplified with:
```suggestion
      if (checkbox instanceof HTMLInputElement) {
          return;
      }
```

Reply:
> It's a bit weird to check the type since the selector is supposed to match what we expect But why not ^^

Aha, I don't like it either, but Eslinter isn't happy without this :(

> However maybe it can be simplified with:

return doesn't work in forEach :stuck_out_tongue_closed_eyes:

Reply:
Oh I see, and is it possible to type the function parameter?

`categoryTree.querySelectorAll(ProductCategoryMap.checkboxInput).forEach((checkbox: HTMLInputElement) => {`